File: swift/examples/contacts_manager/contacts_manager.swift
import Foundation
import Uimd

private let kExpectedLayoutEntries = 11
private let kExpectedFormLayoutEntries = 21
private let kCategoryDefault = "Client"
private let kDataPathEnv = "UI_CONTACTS_MANAGER_DATA_PATH"
private let kCsvColumns = [
    "id",
    "first_name",
    "last_name",
    "email",
    "phone",
    "company",
    "city",
    "category",
    "active",
    "notes",
]

private struct Contact
{
    var id = ""
    var firstName = ""
    var lastName = ""
    var email = ""
    var phone = ""
    var company = ""
    var city = ""
    var category = ""
    var active = true
    var notes = ""
}

private typealias ContactRecord = [String: String]

private func envString(_ name: String) -> String?
{
    let value = ProcessInfo.processInfo.environment[name] ?? ""
    return value.isEmpty ? nil : value
}

private func dataPath() -> String
{
    if let path = envString(kDataPathEnv)
    {
        return path
    }
    return URL(fileURLWithPath: #filePath)
        .deletingLastPathComponent()
        .appendingPathComponent("data")
        .appendingPathComponent("contacts.csv")
        .standardizedFileURL
        .path
}

private func splitCsvLine(_ line: String) -> [String]
{
    var cells: [String] = []
    var cell = ""
    var quoted = false
    for ch in line
    {
        if ch == "\""
        {
            quoted.toggle()
        }
        else if ch == "," && !quoted
        {
            cells.append(cell.trimmingCharacters(in: CharacterSet(charactersIn: "\r")))
            cell = ""
        }
        else
        {
            cell.append(ch)
        }
    }
    cells.append(cell.trimmingCharacters(in: CharacterSet(charactersIn: "\r")))
    return cells
}

private func toBool(_ value: String) -> Bool
{
    let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    return normalized == "1" || normalized == "true" || normalized == "yes" || normalized == "y"
}

private func loadContacts() -> [Contact]
{
    guard let text = try? String(contentsOfFile: dataPath(), encoding: .utf8) else
    {
        return []
    }
    var contacts: [Contact] = []
    for (index, line) in text.components(separatedBy: "\n").enumerated()
    {
        if index == 0 || line.isEmpty
        {
            continue
        }
        let row = splitCsvLine(line)
        if row.count < kCsvColumns.count
        {
            continue
        }
        contacts.append(Contact(
            id: row[0],
            firstName: row[1],
            lastName: row[2],
            email: row[3],
            phone: row[4],
            company: row[5],
            city: row[6],
            category: row[7],
            active: toBool(row[8]),
            notes: row[9]
        ))
    }
    return contacts
}

private func saveContacts(_ contacts: [Contact])
{
    let path = dataPath()
    let parent = URL(fileURLWithPath: path).deletingLastPathComponent().path
    try? FileManager.default.createDirectory(atPath: parent, withIntermediateDirectories: true)
    var lines = [kCsvColumns.joined(separator: ",")]
    for contact in contacts
    {
        lines.append([
            contact.id,
            contact.firstName,
            contact.lastName,
            contact.email,
            contact.phone,
            contact.company,
            contact.city,
            contact.category,
            contact.active ? "true" : "false",
            contact.notes,
        ].joined(separator: ","))
    }
    try? (lines.joined(separator: "\n") + "\n").write(toFile: path, atomically: true, encoding: .utf8)
}

private func contactFromRecord(_ record: ContactRecord) -> Contact
{
    Contact(
        id: record["id"] ?? "",
        firstName: record["first_name"] ?? "",
        lastName: record["last_name"] ?? "",
        email: record["email"] ?? "",
        phone: record["phone"] ?? "",
        company: record["company"] ?? "",
        city: record["city"] ?? "",
        category: record["category"] ?? "",
        active: toBool(record["active"] ?? "true"),
        notes: record["notes"] ?? ""
    )
}

private func recordFromContact(_ contact: Contact) -> ContactRecord
{
    [
        "id": contact.id,
        "first_name": contact.firstName,
        "last_name": contact.lastName,
        "email": contact.email,
        "phone": contact.phone,
        "company": contact.company,
        "city": contact.city,
        "category": contact.category,
        "active": contact.active ? "true" : "false",
        "notes": contact.notes,
    ]
}

private func rowLabel(_ contact: Contact) -> String
{
    (contact.firstName + " " + contact.lastName).trimmingCharacters(in: .whitespacesAndNewlines)
}

private func detailText(_ contact: Contact) -> String
{
    contact.firstName + " " + contact.lastName
        + "\nEmail: " + contact.email
        + "\nPhone: " + contact.phone
        + "\nCompany: " + contact.company
        + "\nCity: " + contact.city
        + "\nCategory: " + contact.category + " (" + (contact.active ? "active" : "inactive") + ")"
        + "\n\n" + contact.notes
}

private final class MessageBoxYesNo: MessageBoxYesNoUI
{
    init(_ header: String, _ text: String)
    {
        super.init()
        dialog_header.setText(header)
        message.setText(text)
    }
}

private final class ContactForm: ContactFormUI
{
    private var contact: ContactRecord
    private(set) var result: ContactRecord = [:]

    init(_ contact: ContactRecord = [:])
    {
        self.contact = contact
        super.init()
        loadContact(contact)
    }

    override func onSaveBtnClick()
    {
        _ = submit()
    }

    override func onCancelBtnClick()
    {
        cancel()
    }

    func submit() -> Bool
    {
        if let error = validationError()
        {
            message.setText(error)
            return false
        }
        var updated = contact
        updated["first_name"] = first_name.value.trimmingCharacters(in: .whitespacesAndNewlines)
        updated["last_name"] = last_name.value.trimmingCharacters(in: .whitespacesAndNewlines)
        updated["email"] = email.value.trimmingCharacters(in: .whitespacesAndNewlines)
        updated["phone"] = phone.value.trimmingCharacters(in: .whitespacesAndNewlines)
        updated["company"] = company.value.trimmingCharacters(in: .whitespacesAndNewlines)
        updated["city"] = city.value.trimmingCharacters(in: .whitespacesAndNewlines)
        updated["category"] = selectedCategory()
        updated["active"] = active.checked ? "true" : "false"
        updated["notes"] = notes.value.trimmingCharacters(in: .whitespacesAndNewlines)
        result = updated
        return true
    }

    func cancel()
    {
        result = [:]
    }

    private func loadContact(_ contact: ContactRecord)
    {
        first_name.setValue(contact["first_name"] ?? "")
        last_name.setValue(contact["last_name"] ?? "")
        email.setValue(contact["email"] ?? "")
        phone.setValue(contact["phone"] ?? "")
        company.setValue(contact["company"] ?? "")
        city.setValue(contact["city"] ?? "")
        category.selectValue(contact["category"] ?? kCategoryDefault)
        active.setChecked(toBool(contact["active"] ?? "true"))
        notes.setValue(contact["notes"] ?? "")
    }

    private func selectedCategory() -> String
    {
        category.selectedIndex >= 0 && category.selectedIndex < category.options.count
            ? category.options[category.selectedIndex]
            : kCategoryDefault
    }

    private func validationError() -> String?
    {
        if first_name.value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        {
            return "First name is required."
        }
        if last_name.value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        {
            return "Last name is required."
        }
        let value = email.value.trimmingCharacters(in: .whitespacesAndNewlines)
        if value.isEmpty
        {
            return "Email is required."
        }
        guard let at = value.firstIndex(of: "@"), at != value.startIndex else
        {
            return "Email must look like name@example.com."
        }
        let domain = value[value.index(after: at)...]
        if !domain.contains(".")
        {
            return "Email must look like name@example.com."
        }
        return nil
    }
}

public final class ContactsManagerApp: ContactsManagerUI
{
    private let modalStack = GeneratedWindowStack()
    private var contactsModel: [Contact] = []
    private var form: ContactForm?
    private var deleteDialog: MessageBoxYesNo?
    private var deleteContactId = ""
    private var finished = false
    private var dirty = false

    public override init()
    {
        super.init()
        reload()
    }

    public override func runtimeOptions() -> GeneratedWindowRuntimeOptions
    {
        var options = super.runtimeOptions()
        options.windowStack = modalStack
        return options
    }

    public override func shouldClose() -> Bool
    {
        finished
    }

    override public func onContactsSelectionChange(_ value: [String])
    {
        _ = value
        refreshDetail()
    }

    override public func onAddBtnClick()
    {
        addContact()
    }

    override public func onEditBtnClick()
    {
        editSelected()
    }

    override public func onDeleteBtnClick()
    {
        confirmDeleteSelected()
    }

    override public func onSaveBtnClick()
    {
        save()
    }

    override public func onReloadBtnClick()
    {
        reload()
    }

    override public func onCloseBtnClick()
    {
        finished = true
    }

    func reload()
    {
        contactsModel = loadContacts()
        dirty = false
        refreshList(selectedId())
        setStatus("Loaded " + String(contactsModel.count) + " contacts.")
    }

    func save()
    {
        saveContacts(contactsModel)
        dirty = false
        setStatus("Saved " + String(contactsModel.count) + " contacts.")
    }

    func refreshDetail()
    {
        guard let contact = selectedContact() else
        {
            detail.setText("No contact selected.")
            return
        }
        detail.setText(detailText(contact))
    }

    func addContact()
    {
        openForm(nil)
    }

    func editSelected()
    {
        guard let contact = selectedContact() else
        {
            setStatus("Select a contact to edit.")
            return
        }
        openForm(contact)
    }

    func confirmDeleteSelected()
    {
        guard let contact = selectedContact() else
        {
            setStatus("Select a contact to delete.")
            return
        }
        deleteContactId = contact.id
        deleteDialog = MessageBoxYesNo("Delete Contact", "Delete " + rowLabel(contact) + "?")
        guard let deleteDialog else
        {
            return
        }
        var frame = GeneratedWindowFrameOptions()
        frame.className = "MessageBoxYesNo"
        frame.initialFocusName = "no_btn"
        frame.onButton = { [weak self] name in
            if name == "yes_btn"
            {
                self?.closeDeleteDialog("yes")
                return true
            }
            if name == "no_btn"
            {
                self?.closeDeleteDialog("no")
                return true
            }
            return false
        }
        frame.onKey = { [weak self] key in
            if key == "Escape"
            {
                self?.closeDeleteDialog("no")
                return true
            }
            return false
        }
        modalStack.push(deleteDialog, frame)
    }

    private func onFormClosed(_ result: ContactRecord)
    {
        if result.isEmpty
        {
            setStatus("Edit cancelled.")
            return
        }
        if (result["id"] ?? "").isEmpty
        {
            var updated = result
            updated["id"] = String(nextId())
            contactsModel.append(contactFromRecord(updated))
            setStatus("Contact added.")
            dirty = true
            refreshList(updated["id"] ?? "")
            return
        }
        for index in contactsModel.indices
        {
            if contactsModel[index].id == result["id"]
            {
                contactsModel[index] = contactFromRecord(result)
                setStatus("Contact updated.")
                dirty = true
                refreshList(result["id"] ?? "")
                return
            }
        }
        setStatus("Contact no longer exists.")
    }

    func addContactForTest(firstName: String, lastName: String, email: String)
    {
        let form = ContactForm()
        form.first_name.setValue(firstName)
        form.last_name.setValue(lastName)
        form.email.setValue(email)
        if form.submit()
        {
            onFormClosed(form.result)
        }
    }

    func editSelectedForTest(email: String)
    {
        guard let contact = selectedContact() else
        {
            setStatus("Select a contact to edit.")
            return
        }
        let form = ContactForm(recordFromContact(contact))
        form.email.setValue(email)
        if form.submit()
        {
            onFormClosed(form.result)
        }
    }

    func deleteConfirmed(contactId: String, result: String)
    {
        if result != "yes"
        {
            setStatus("Delete cancelled.")
            return
        }
        let before = contactsModel.count
        contactsModel.removeAll { $0.id == contactId }
        if contactsModel.count == before
        {
            setStatus("Contact no longer exists.")
            return
        }
        dirty = true
        refreshList("")
        setStatus("Contact deleted.")
    }

    private func openForm(_ contact: Contact?)
    {
        form = ContactForm(contact.map(recordFromContact) ?? [:])
        guard let form else
        {
            return
        }
        var frame = GeneratedWindowFrameOptions()
        frame.className = "ContactForm"
        frame.onButton = { [weak self] name in
            if name == "save_btn"
            {
                self?.submitForm()
                return true
            }
            if name == "cancel_btn"
            {
                self?.cancelForm()
                return true
            }
            return false
        }
        modalStack.push(form, frame)
    }

    private func submitForm()
    {
        guard let form else
        {
            return
        }
        if !form.submit()
        {
            return
        }
        let result = form.result
        modalStack.remove(form)
        self.form = nil
        onFormClosed(result)
    }

    private func cancelForm()
    {
        guard let form else
        {
            return
        }
        form.cancel()
        modalStack.remove(form)
        self.form = nil
        onFormClosed([:])
    }

    private func closeDeleteDialog(_ result: String)
    {
        if let deleteDialog
        {
            modalStack.remove(deleteDialog)
        }
        deleteDialog = nil
        let contactId = deleteContactId
        deleteContactId = ""
        deleteConfirmed(contactId: contactId, result: result)
    }

    private func refreshList(_ selectedContactId: String)
    {
        let labels = contactsModel.map(rowLabel)
        header.setText("Contacts Manager - " + String(labels.count) + " records")
        contacts.options = labels
        if labels.isEmpty
        {
            contacts.selectedIndex = -1
            contacts.selectedValues = []
            refreshDetail()
            return
        }
        var selectedIndex = 0
        if !selectedContactId.isEmpty,
           let index = contactsModel.firstIndex(where: { $0.id == selectedContactId })
        {
            selectedIndex = index
        }
        contacts.selectedIndex = selectedIndex
        contacts.selectedValues = [labels[selectedIndex]]
        refreshDetail()
    }

    private func selectedContact() -> Contact?
    {
        if contactsModel.isEmpty
        {
            return nil
        }
        let index = min(max(0, contacts.selectedIndex), contactsModel.count - 1)
        return contactsModel[index]
    }

    private func selectedId() -> String
    {
        selectedContact()?.id ?? ""
    }

    private func nextId() -> Int
    {
        contactsModel.reduce(1) { result, contact in
            max(result, (Int(contact.id) ?? 0) + 1)
        }
    }

    private func setStatus(_ text: String)
    {
        status.setText(text + (dirty ? " *" : ""))
    }
}

private func smoke()
{
    let app = ContactsManagerApp()
    let form = ContactForm()
    precondition(app.windowTitle == "Contacts Manager")
    precondition(app.layout.count == kExpectedLayoutEntries)
    precondition(form.layout.count == kExpectedFormLayoutEntries)
    precondition(app.contacts != nil)
    precondition(app.detail != nil)
    precondition(app.add_btn != nil)
    precondition(app.close_btn != nil)
    precondition(!app.contacts.options.isEmpty)
    precondition(app.header.text.contains("records"))
    precondition(!app.detail.text.contains("\r"))
    let contentSize = app.generatedContentSize()
    _ = app.renderContent(size: contentSize, focusedName: nil, editMode: false)
    precondition(app.contacts.frame.height == app.contacts.options.count)
    let widerContentSize = Size(width: contentSize.width + app.close_btn.frame.width, height: contentSize.height)
    _ = app.renderContent(size: widerContentSize, focusedName: nil, editMode: false)
    precondition(app.close_btn.frame.col + app.close_btn.frame.width == widerContentSize.width)
}

private func logicTest()
{
    let app = ContactsManagerApp()
    let initialSize = app.contacts.options.count
    precondition(initialSize > 0)
    app.contacts.selectedIndex = 0
    app.refreshDetail()
    precondition(app.detail.text.contains("Email:"))
    app.addContactForTest(firstName: "New", lastName: "Contact", email: "new@example.com")
    precondition(app.contacts.options.count == initialSize + 1)
    precondition(app.status.text == "Contact added.")
    app.editSelectedForTest(email: "updated@example.com")
    precondition(app.status.text == "Contact updated. *")
    let editedSize = app.contacts.options.count
    app.deleteConfirmed(contactId: "999999", result: "no")
    precondition(app.contacts.options.count == editedSize)
}

@main
enum Program
{
    static func main()
    {
        let code = GeneratedWindowRuntime.runGeneratedAppMain
        {
            if CommandLine.arguments.contains("--smoke")
            {
                smoke()
                return 0
            }
            if CommandLine.arguments.contains("--logic-test")
            {
                logicTest()
                return 0
            }
            let app = ContactsManagerApp()
            return GeneratedWindowRuntime.runGeneratedWindow(app, app.runtimeOptions(), CommandLine.arguments)
        }
        Foundation.exit(Int32(code))
    }
}

File: go/examples/contacts_manager/contacts_manager.go
package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	contactform "uimd_examples/contacts_manager/contact_form"

	uimd "uimd"
)

const contactsDataPathEnv = "UI_CONTACTS_MANAGER_DATA_PATH"

var contactCsvColumns = []string{
	"id", "first_name", "last_name", "email", "phone", "company", "city", "category", "active", "notes",
}

type Contact struct {
	ID        string
	FirstName string
	LastName  string
	Email     string
	Phone     string
	Company   string
	City      string
	Category  string
	Active    bool
	Notes     string
}

type ContactsManagerApp struct {
	*ContactsManagerUI
	stack           uimd.GeneratedWindowStack
	contacts        []Contact
	form            *contactform.ContactForm
	deleteDialog    *uimd.MessageBoxYesNo
	deleteContactID string
	finished        bool
	dirty           bool
}

func NewContactsManagerApp() *ContactsManagerApp {
	app := &ContactsManagerApp{ContactsManagerUI: NewContactsManagerUI()}
	app.SetEventHandler(app)
	app.Reload()
	return app
}

func (app *ContactsManagerApp) RuntimeOptions() uimd.GeneratedWindowRuntimeOptions {
	options := app.ContactsManagerUI.RuntimeOptions()
	options.WindowStack = &app.stack
	options.ShouldClose = app.ShouldClose
	return options
}

func (app *ContactsManagerApp) ShouldClose() bool {
	return app.finished || app.GeneratedWindowBase.ShouldClose()
}

func (app *ContactsManagerApp) OnContactsSelectionChange([]string) {
	app.RefreshDetail()
}

func (app *ContactsManagerApp) OnReloadBtnClick() {
	app.Reload()
}

func (app *ContactsManagerApp) OnAddBtnClick() {
	app.AddContact()
}

func (app *ContactsManagerApp) OnEditBtnClick() {
	app.EditSelected()
}

func (app *ContactsManagerApp) OnSaveBtnClick() {
	app.Save()
}

func (app *ContactsManagerApp) OnDeleteBtnClick() {
	app.ConfirmDeleteSelected()
}

func (app *ContactsManagerApp) OnCloseBtnClick() {
	app.finished = true
	app.RequestClose()
}

func (app *ContactsManagerApp) Reload() {
	app.contacts = loadContacts()
	app.dirty = false
	app.refreshList(app.selectedID())
	app.setStatus(fmt.Sprintf("Loaded %d contacts.", len(app.contacts)))
}

func (app *ContactsManagerApp) Save() {
	saveContacts(app.contacts)
	app.dirty = false
	app.setStatus(fmt.Sprintf("Saved %d contacts.", len(app.contacts)))
}

func (app *ContactsManagerApp) RefreshDetail() {
	contact := app.selectedContact()
	if contact == nil {
		app.Detail.SetText("No contact selected.")
		return
	}
	app.Detail.SetText(detailText(*contact))
}

func (app *ContactsManagerApp) AddContact() {
	app.openForm(nil)
}

func (app *ContactsManagerApp) EditSelected() {
	contact := app.selectedContact()
	if contact == nil {
		app.setStatus("Select a contact to edit.")
		return
	}
	app.openForm(contact)
}

func (app *ContactsManagerApp) ConfirmDeleteSelected() {
	contact := app.selectedContact()
	if contact == nil {
		app.setStatus("Select a contact to delete.")
		return
	}
	app.deleteContactID = contact.ID
	app.deleteDialog = uimd.NewMessageBoxYesNo("Delete Contact", "Delete "+rowLabel(*contact)+"?")
	frame := uimd.GeneratedWindowFrameOptions{
		ClassName:        "MessageBoxYesNo",
		InitialFocusName: "no_btn",
		OnButton: func(name string) {
			switch name {
			case "yes_btn":
				app.closeDeleteDialog("yes")
			case "no_btn":
				app.closeDeleteDialog("no")
			}
		},
		OnKey: func(key string) bool {
			if key == "Escape" {
				app.closeDeleteDialog("no")
				return true
			}
			return false
		},
	}
	app.stack.Push(app.deleteDialog, frame)
}

func (app *ContactsManagerApp) OnFormClosed(result contactform.ContactRecord) {
	if len(result) == 0 {
		app.setStatus("Edit cancelled.")
		return
	}
	if result["id"] == "" {
		result["id"] = strconv.Itoa(app.nextID())
		app.contacts = append(app.contacts, contactFromRecord(result))
		app.setStatus("Contact added.")
		app.dirty = true
		app.refreshList(result["id"])
		return
	}
	for index := range app.contacts {
		if app.contacts[index].ID == result["id"] {
			app.contacts[index] = contactFromRecord(result)
			app.setStatus("Contact updated.")
			app.dirty = true
			app.refreshList(result["id"])
			return
		}
	}
	app.setStatus("Contact no longer exists.")
}

func (app *ContactsManagerApp) AddContactForTest(firstName string, lastName string, email string) {
	form := contactform.NewContactForm(contactform.ContactRecord{}, func(result contactform.ContactRecord) {
		app.OnFormClosed(result)
	})
	form.FirstName.SetValue(firstName)
	form.LastName.SetValue(lastName)
	form.Email.SetValue(email)
	form.Submit()
}

func (app *ContactsManagerApp) EditSelectedForTest(email string) {
	contact := app.selectedContact()
	if contact == nil {
		app.setStatus("Select a contact to edit.")
		return
	}
	form := contactform.NewContactForm(recordFromContact(*contact), func(result contactform.ContactRecord) {
		app.OnFormClosed(result)
	})
	form.Email.SetValue(email)
	form.Submit()
}

func (app *ContactsManagerApp) DeleteConfirmed(contactID string, result string) {
	if result != "yes" {
		app.setStatus("Delete cancelled.")
		return
	}
	before := len(app.contacts)
	filtered := app.contacts[:0]
	for _, contact := range app.contacts {
		if contact.ID != contactID {
			filtered = append(filtered, contact)
		}
	}
	app.contacts = filtered
	if len(app.contacts) == before {
		app.setStatus("Contact no longer exists.")
		return
	}
	app.dirty = true
	app.refreshList("")
	app.setStatus("Contact deleted.")
}

func (app *ContactsManagerApp) openForm(contact *Contact) {
	record := contactform.ContactRecord{}
	if contact != nil {
		record = recordFromContact(*contact)
	}
	app.form = contactform.NewContactForm(record, nil)
	frame := uimd.GeneratedWindowFrameOptions{
		ClassName: "ContactForm",
		OnButton: func(name string) {
			switch name {
			case "save_btn":
				app.submitForm()
			case "cancel_btn":
				app.cancelForm()
			}
		},
	}
	app.stack.Push(app.form.GeneratedWindowBase, frame)
}

func (app *ContactsManagerApp) submitForm() {
	if app.form == nil {
		return
	}
	if !app.form.Submit() {
		return
	}
	result := app.form.Result()
	app.stack.Remove(app.form.GeneratedWindowBase)
	app.form = nil
	app.OnFormClosed(result)
}

func (app *ContactsManagerApp) cancelForm() {
	if app.form == nil {
		return
	}
	app.form.Cancel()
	app.stack.Remove(app.form.GeneratedWindowBase)
	app.form = nil
	app.OnFormClosed(contactform.ContactRecord{})
}

func (app *ContactsManagerApp) closeDeleteDialog(result string) {
	if app.deleteDialog != nil {
		app.stack.Remove(app.deleteDialog)
	}
	app.deleteDialog = nil
	contactID := app.deleteContactID
	app.deleteContactID = ""
	app.DeleteConfirmed(contactID, result)
}

func (app *ContactsManagerApp) refreshList(selectedContactID string) {
	labels := make([]string, 0, len(app.contacts))
	for _, contact := range app.contacts {
		labels = append(labels, rowLabel(contact))
	}
	app.Header.SetText(fmt.Sprintf("Contacts Manager - %d records", len(labels)))
	app.Contacts.SetOptions(labels)
	selectedIndex := 0
	if selectedContactID != "" {
		for index, contact := range app.contacts {
			if contact.ID == selectedContactID {
				selectedIndex = index
				break
			}
		}
	}
	app.Contacts.SetSelectedIndex(selectedIndex)
	app.RefreshDetail()
}

func (app *ContactsManagerApp) selectedContact() *Contact {
	if len(app.contacts) == 0 {
		return nil
	}
	index := app.Contacts.SelectedIndex
	if index < 0 {
		index = 0
	}
	if index >= len(app.contacts) {
		index = len(app.contacts) - 1
	}
	return &app.contacts[index]
}

func (app *ContactsManagerApp) selectedID() string {
	contact := app.selectedContact()
	if contact == nil {
		return ""
	}
	return contact.ID
}

func (app *ContactsManagerApp) nextID() int {
	result := 1
	for _, contact := range app.contacts {
		value, err := strconv.Atoi(contact.ID)
		if err == nil && value >= result {
			result = value + 1
		}
	}
	return result
}

func (app *ContactsManagerApp) setStatus(text string) {
	if app.dirty {
		text += " *"
	}
	app.Status.SetText(text)
}

func dataPath() string {
	if value := os.Getenv(contactsDataPathEnv); value != "" {
		return value
	}
	return filepath.Join("data", "contacts.csv")
}

func loadContacts() []Contact {
	file, err := os.Open(dataPath())
	if err != nil {
		return []Contact{}
	}
	defer file.Close()
	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil || len(records) <= 1 {
		return []Contact{}
	}
	contacts := make([]Contact, 0, len(records)-1)
	for _, row := range records[1:] {
		if len(row) < len(contactCsvColumns) {
			continue
		}
		contacts = append(contacts, Contact{
			ID:        row[0],
			FirstName: row[1],
			LastName:  row[2],
			Email:     row[3],
			Phone:     row[4],
			Company:   row[5],
			City:      row[6],
			Category:  row[7],
			Active:    contactformBool(row[8]),
			Notes:     row[9],
		})
	}
	return contacts
}

func saveContacts(contacts []Contact) {
	_ = os.MkdirAll(filepath.Dir(dataPath()), 0o755)
	file, err := os.Create(dataPath())
	if err != nil {
		return
	}
	defer file.Close()
	writer := csv.NewWriter(file)
	_ = writer.Write(contactCsvColumns)
	for _, contact := range contacts {
		active := "false"
		if contact.Active {
			active = "true"
		}
		_ = writer.Write([]string{
			contact.ID, contact.FirstName, contact.LastName, contact.Email, contact.Phone,
			contact.Company, contact.City, contact.Category, active, contact.Notes,
		})
	}
	writer.Flush()
}

func contactFromRecord(record contactform.ContactRecord) Contact {
	return Contact{
		ID:        record["id"],
		FirstName: record["first_name"],
		LastName:  record["last_name"],
		Email:     record["email"],
		Phone:     record["phone"],
		Company:   record["company"],
		City:      record["city"],
		Category:  record["category"],
		Active:    contactformBool(record["active"]),
		Notes:     record["notes"],
	}
}

func recordFromContact(contact Contact) contactform.ContactRecord {
	active := "false"
	if contact.Active {
		active = "true"
	}
	return contactform.ContactRecord{
		"id":         contact.ID,
		"first_name": contact.FirstName,
		"last_name":  contact.LastName,
		"email":      contact.Email,
		"phone":      contact.Phone,
		"company":    contact.Company,
		"city":       contact.City,
		"category":   contact.Category,
		"active":     active,
		"notes":      contact.Notes,
	}
}

func rowLabel(contact Contact) string {
	return strings.TrimSpace(contact.FirstName + " " + contact.LastName)
}

func detailText(contact Contact) string {
	active := "inactive"
	if contact.Active {
		active = "active"
	}
	return contact.FirstName + " " + contact.LastName +
		"\nEmail: " + contact.Email +
		"\nPhone: " + contact.Phone +
		"\nCompany: " + contact.Company +
		"\nCity: " + contact.City +
		"\nCategory: " + contact.Category + " (" + active + ")" +
		"\n\n" + contact.Notes
}

func contactformBool(value string) bool {
	switch strings.ToLower(value) {
	case "1", "true", "yes", "y":
		return true
	default:
		return false
	}
}

func smoke() {
	app := NewContactsManagerApp()
	if app.Title() != "Contacts Manager" {
		panic("unexpected title")
	}
	if app.Contacts == nil || app.Detail == nil || app.AddBtn == nil || app.CloseBtn == nil {
		panic("missing contacts manager elements")
	}
	if len(app.Contacts.Options) == 0 {
		panic("contacts did not load")
	}
}

func logicTest() {
	app := NewContactsManagerApp()
	initialSize := len(app.Contacts.Options)
	if initialSize == 0 {
		panic("contacts did not load")
	}
	app.Contacts.SetSelectedIndex(0)
	app.RefreshDetail()
	if !strings.Contains(app.Detail.Text, "Email:") {
		panic("detail missing email")
	}
	app.AddContactForTest("New", "Contact", "new@example.com")
	if len(app.Contacts.Options) != initialSize+1 {
		panic("contact not added")
	}
	app.EditSelectedForTest("updated@example.com")
	if app.Status.Text != "Contact updated. *" {
		panic("contact not updated")
	}
	editedSize := len(app.Contacts.Options)
	app.DeleteConfirmed("999999", "no")
	if len(app.Contacts.Options) != editedSize {
		panic("delete cancel changed contacts")
	}
}

func main() {
	code := uimd.RunGeneratedAppMain(func() int {
		if len(os.Args) > 1 {
			switch os.Args[1] {
			case "--smoke":
				smoke()
				return 0
			case "--logic-test":
				logicTest()
				return 0
			}
		}
		app := NewContactsManagerApp()
		return uimd.RunGeneratedWindow(app, app.RuntimeOptions(), os.Args)
	})
	os.Exit(code)
}
